scionlab/models/user_as.py

```diff
@@ -192,100 +127,208 @@ class Meta:
+            ap.trigger_deployment()
+
+    def _create_or_update_attachment(self, att_conf):
+        """
+        Proxy function that either calls create or update over a UserAS
+        attachment point configuration, based on the presence of a link object
+        """
+        if att_conf.link is None:
+            assert att_conf.active, "Cannot deactivate a non existing Link"
+            return self._create_attachment(att_conf)
+        else:
+            return self._update_attachment(att_conf)
+
+    def _create_attachment(self, att_conf):
+        """
+        Attach a UserAS to the specified attachment point creating the Link and the Interfaces,
+        and if needed a VPNClient. The new `Link` is stored in `att_conf.link` as side effect
+        :param 'AttachmentPointConf' att_conf:
+        """
+        ap = att_conf.attachment_point
+        if self.isd != ap.AS.isd:
+            self._change_isd(ap.AS.isd)
+        if att_conf.use_vpn:
+            iface_ap, iface_client = self._create_or_update_vpn_connection(att_conf)
+        else:
+            _bind_ip = att_conf.bind_ip
+            if self.installation_type == UserAS.VM:
+                _bind_ip = _VAGRANT_VM_LOCAL_IP
+            iface_client = Interface.objects.create(
+                border_router=BorderRouter.objects.first_or_create(self.host),
+                public_ip=att_conf.public_ip,
+                public_port=att_conf.public_port,
+                bind_ip=_bind_ip,
+                bind_port=att_conf.bind_port
+            )
+            iface_ap = Interface.objects.create(
+                border_router=ap.get_border_router_for_useras_interface(),
+            )
+
+        link = Link.objects.create(
+            type=Link.PROVIDER,
+            interfaceA=iface_ap,
+            interfaceB=iface_client
+        )
+        self.save()
+        att_conf.link = link
+
+    def _update_attachment(self, att_conf):
+        """
+        Update a Link and Interfaces(A and B) involved in a UserAS attachment
+        :param 'AttachmentPointConf' att_conf:
+        """
+        ap = att_conf.attachment_point
+        if att_conf.active and self.isd != ap.AS.isd:
+            self._change_isd(ap.AS.isd)
+
+        att_conf.link.update(active=att_conf.active)
+        att_conf.link.save()
+        if ap.vpn is None:
+            assert not att_conf.use_vpn, "AttachmentPoint doesn't support VPN connections"
+            # We simply cannot have a VPN connection if the ap does not support VPN
+            vpn_client = None
+        else:
+            vpn_client = VPNClient.objects.filter(host=self.host, vpn=ap.vpn).first()
+
+        if att_conf.use_vpn:
+            # Create or Update the vpn_client accordingly
+            iface_ap, iface_client = self._create_or_update_vpn_connection(att_conf)
+        else:
+            if vpn_client is not None and vpn_client.active:
+                # Deactivate the existing vpn_client
+                vpn_client.active = False
```

Comment:
As discussed: this needs to consider whether _any_ of the related `att_conf`'s has vpn enabled.